packages/jsii/lib/transforms/deprecated-warnings-transformer.ts

```diff
@@ -0,0 +1,413 @@
+    if (declaration != null && declaration !== node) {
+      if (isDeprecated(declaration)) {
+        warnings.push(
+          getWarning(declaration, ElementType.PARAMETER, this.moduleName),
+        );
+      }
+      if (
+        ts.isInterfaceDeclaration(declaration) ||
+        ts.isClassDeclaration(node)
+      ) {
+        newBatch = type.getProperties().map((p) => p.declarations[0]);
+      }
+      if (ts.isEnumDeclaration(declaration)) {
+        newBatch = [...declaration.members];
+      }
+    }
+    return this.getParameterWarnings(
+      toProcess.slice(1).concat(newBatch),
+      result.concat(warnings),
+    );
+  }
+
+  private createWarningStatements(): ts.Statement[] {
+    return [...this.warnings].flatMap((w) => this.createWarningStatement(w));
+  }
+
+  // TODO Rename this method
+  private findDeprecatedInTheTypeHierarchy(
+    toProcess: ts.Type[],
+  ): ts.Type | undefined {
+    if (toProcess.length === 0) {
+      return undefined;
+    }
+
+    const type = toProcess[0];
+    if (this.deprecatedTypes.includes(type)) {
+      return type;
+    }
+
+    const node = this.index.get(
+      this.typeChecker.getFullyQualifiedName(type.symbol),
+    );
+
+    // This type was not declared in the JSII assembly. Skipping.
+    if (node == null) {
+      return undefined;
+    }
+
+    if (!ts.isClassDeclaration(node) && !ts.isInterfaceDeclaration(node)) {
+      throw new Error('Node should be a class or interface declaration.');
+    }
+
+    const newBatch = toProcess.slice(1);
+    const heritageClauses = node.heritageClauses ?? [];
+    for (const clause of heritageClauses) {
+      for (const expression of clause.types) {
+        const type = this.typeChecker.getTypeAtLocation(expression);
+        if (!toProcess.includes(type)) {
+          newBatch.push(type);
+        }
+      }
+    }
+
+    return this.findDeprecatedInTheTypeHierarchy(newBatch);
+  }
+
+  private createWarningStatement(warning: Warning): ts.Statement[] {
+    const message = `${warning?.elementName} is deprecated.\n  ${warning?.message}\n  This API will be removed in the net major release.`;
+    // TODO Are random numbers a good idea?
+    const functionName = `printWarning${getRandomInt(
+      0,
+      Number.MAX_SAFE_INTEGER,
+    )}`;
+
+    // TODO There must be a simpler way...
+    return [
+      // TODO Instead of declaring the function everywhere it's used, should we create a single declaration somewhere and reuse it?
+      //  If so, where should this declaration go?
+      ts.createFunctionDeclaration(
+        undefined,
+        undefined,
+        undefined,
+        functionName,
+        [],
+        [],
+        undefined,
+        ts.createBlock(
+          [
+            ts.createExpressionStatement(
+              ts.createAssignment(
+                ts.createIdentifier('const deprecated'),
+                ts.createIdentifier('process.env.DEPRECATED'),
```

Comment:
That environment variable name is way too generic IMHO. Should likely be `JSII_DEPRECATIONS` or something along those lines.

Reply:
Makes sense. I'll rename it.